File: robert-crypto-grpc-server/src/test/java/fr/gouv/stopc/robertserver/crypto/test/PostgreSqlManager.java
package fr.gouv.stopc.robertserver.crypto.test;

import com.google.protobuf.ByteString;
import fr.gouv.stopc.robert.server.crypto.structure.impl.CryptoAESGCM;
import fr.gouv.stopc.robert.server.crypto.structure.impl.CryptoHMACSHA256;
import lombok.SneakyThrows;
import org.assertj.core.api.ListAssert;
import org.flywaydb.core.Flyway;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.test.context.TestContext;
import org.springframework.test.context.TestExecutionListener;
import org.testcontainers.containers.JdbcDatabaseContainer;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.utility.DockerImageName;

import java.security.SecureRandom;
import java.util.Base64;
import java.util.Map;

import static fr.gouv.stopc.robertserver.crypto.test.KeystoreManager.cipherForStoredKey;
import static org.assertj.core.api.Assertions.assertThat;

/**
 * A {@link TestExecutionListener} to start a PostgreSQL container to be used as
 * a dependency for SpringBootTests.
 * <p>
 * It starts a postgresql container statically and export required system
 * properties to override Spring application context configuration.
 */
public class PostgreSqlManager implements TestExecutionListener {

    private static final JdbcDatabaseContainer POSTGRE = new PostgreSQLContainer<>(
            DockerImageName.parse("postgres:9.6")
    );

    private static JdbcTemplate jdbcTemplate;

    static {
        POSTGRE.start();
        System.setProperty("spring.datasource.url", POSTGRE.getJdbcUrl());
        System.setProperty("spring.datasource.username", POSTGRE.getUsername());
        System.setProperty("spring.datasource.password", POSTGRE.getPassword());
    }

    public static ListAssert<Map<String, Object>> assertThatAllIdentities() {
        return assertThat(jdbcTemplate.queryForList("select * from identity"));
    }

    @Override
    public void beforeTestClass(final TestContext testContext) {
        jdbcTemplate = testContext.getApplicationContext()
                .getBean(JdbcTemplate.class);
    }

    @Override
    public void beforeTestMethod(TestContext testContext) {
        jdbcTemplate.execute("DROP SCHEMA public CASCADE ;");
        jdbcTemplate.execute("CREATE SCHEMA public ;");
        testContext.getApplicationContext()
                .getBean(Flyway.class)
                .migrate();
    }

    @SneakyThrows
    public static void givenIdentityExistsForIdA(String base64EncodedIdA) {
        final var keyForMac = new byte[32];
        final var keyForTuples = new byte[32];
        final var random = new SecureRandom();
        random.nextBytes(keyForMac);
        random.nextBytes(keyForTuples);
        final var encryptedKeyForMac = cipherForStoredKey().encrypt(keyForMac);
        final var encryptedKeyForTuples = cipherForStoredKey().encrypt(keyForTuples);
        jdbcTemplate.update(
                "insert into identity(id, idA, key_for_mac, key_for_tuples, creation_time, last_update)" +
                        " values ((select nextval('identity_id_seq ')), ?, ?, ?, now(), now())",
                base64EncodedIdA,
                Base64.getEncoder().encodeToString(encryptedKeyForMac),
                Base64.getEncoder().encodeToString(encryptedKeyForTuples)
        );
    }

    public static void givenIdentityDoesntExistForIdA(String base64EncodedIdA) {
        jdbcTemplate.update("delete from identity where idA = ?", base64EncodedIdA);
    }

    public static CryptoAESGCM getCipherForTuples(final ByteString idA) {
        final var base64EncodedIdA = Base64.getEncoder().encodeToString(idA.toByteArray());
        return findKeyForTuplesByIdA(base64EncodedIdA);
    }

    @SneakyThrows
    public static CryptoAESGCM findKeyForTuplesByIdA(final String base64EncodedIdA) {
        final var base64Key = jdbcTemplate
                .queryForObject("select key_for_tuples from IDENTITY where idA = ?", String.class, base64EncodedIdA);
        final var encryptedKey = Base64.getDecoder().decode(base64Key);
        final var key = cipherForStoredKey().decrypt(encryptedKey);
        return new CryptoAESGCM(key);
    }

    @SneakyThrows
    public static CryptoHMACSHA256 getCipherForMac(final String base64EncodedIdA) {
        final var base64Key = jdbcTemplate
                .queryForObject("select key_for_mac from identity where idA = ?", String.class, base64EncodedIdA);
        final var encryptedKey = Base64.getDecoder().decode(base64Key);
        final var keyForMac = cipherForStoredKey().decrypt(encryptedKey);
        return new CryptoHMACSHA256(keyForMac);
    }
}

File: robert-crypto-grpc-server/src/test/kotlin/fr/gouv/stopc/robertserver/crypto/test/PostgresqlManager.kt
package fr.gouv.stopc.robertserver.crypto.test

import com.google.protobuf.ByteString
import fr.gouv.stopc.robert.server.crypto.structure.impl.CryptoAESGCM
import fr.gouv.stopc.robert.server.crypto.structure.impl.CryptoHMACSHA256
import fr.gouv.stopc.robertserver.common.base64Decode
import fr.gouv.stopc.robertserver.common.base64Encode
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.ListAssert
import org.flywaydb.core.Flyway
import org.springframework.context.ApplicationContextInitializer
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.test.context.TestContext
import org.springframework.test.context.TestExecutionListener
import org.springframework.test.context.support.TestPropertySourceUtils
import org.testcontainers.containers.PostgreSQLContainer
import org.testcontainers.utility.DockerImageName
import kotlin.random.Random

private lateinit var jdbcTemplate: JdbcTemplate

/**
 * A [TestExecutionListener] to start a PostgreSQL container to be used as a dependency for SpringBootTests.
 *
 * It starts a postgresql container and export required system properties to override Spring application context configuration.
 */
class PostgreSqlManager : TestExecutionListener, ApplicationContextInitializer<ConfigurableApplicationContext> {

    override fun initialize(configurableApplicationContext: ConfigurableApplicationContext) {
        PostgreSQLContainer<Nothing>(DockerImageName.parse("postgres:13.8")).apply {
            start()
            TestPropertySourceUtils.addInlinedPropertiesToEnvironment(
                configurableApplicationContext,
                "spring.datasource.url=$jdbcUrl",
                "spring.datasource.username=$username",
                "spring.datasource.password=$password"
            )
        }
    }

    override fun beforeTestClass(testContext: TestContext) {
        jdbcTemplate = testContext.applicationContext
            .getBean(JdbcTemplate::class.java)
    }

    override fun beforeTestMethod(testContext: TestContext) {
        jdbcTemplate.execute("DROP SCHEMA public CASCADE ;")
        jdbcTemplate.execute("CREATE SCHEMA public ;")
        testContext.applicationContext
            .getBean(Flyway::class.java)
            .migrate()
    }
}

/**
 * Verify existing application identities stored in database.
 */
fun assertThatAllIdentities(): ListAssert<Map<String, Any>> =
    assertThat(jdbcTemplate.queryForList("select * from identity"))

/**
 * Creates an `idA` in the database with two random keys for MAC and TUPLES computation.
 *
 * Generated keys can be retrieved using [getCipherForTuples] and [getCipherForMac].
 */
fun givenIdentityExistsForIdA(base64EncodedIdA: String) {
    val keyForMac = Random.nextBytes(32)
    val keyForTuples = Random.nextBytes(32)
    val encryptedKeyForMac = cipherForStoredKey().encrypt(keyForMac)
    val encryptedKeyForTuples = cipherForStoredKey().encrypt(keyForTuples)
    jdbcTemplate.update(
        """
        insert into identity(id, idA, key_for_mac, key_for_tuples, creation_time, last_update)
            values ((select nextval('identity_id_seq ')), ?, ?, ?, now(), now())
            
        """.trimIndent(),
        base64EncodedIdA,
        encryptedKeyForMac.base64Encode(),
        encryptedKeyForTuples.base64Encode()
    )
}

/**
 * Ensure an `idA` doesn't exist, delete it if necessary.
 */
fun givenIdentityDoesntExistForIdA(base64EncodedIdA: String) {
    jdbcTemplate.update("delete from identity where idA = ?", base64EncodedIdA)
}

/**
 * Returns the cipher to encrypt/decrypt TUPLES for the given `idA`.
 */
fun getCipherForTuples(idA: ByteString): CryptoAESGCM = getCipherForTuples(idA.toByteArray().base64Encode())

/**
 * Returns the cipher to encrypt/decrypt TUPLES for the given `idA`.
 */
fun getCipherForTuples(base64EncodedIdA: String): CryptoAESGCM {
    val base64Key = jdbcTemplate
        .queryForObject("select key_for_tuples from IDENTITY where idA = ?", String::class.java, base64EncodedIdA)
    val key = cipherForStoredKey().decrypt(base64Key.base64Decode())
    return CryptoAESGCM(key)
}

/**
 * Returns the cipher to encrypt/decrypt MACs for the given `idA`.
 */
fun getCipherForMac(base64EncodedIdA: String): CryptoHMACSHA256 {
    val base64Key = jdbcTemplate
        .queryForObject("select key_for_mac from identity where idA = ?", String::class.java, base64EncodedIdA)
    val keyForMac = cipherForStoredKey().decrypt(base64Key.base64Decode())
    return CryptoHMACSHA256(keyForMac)
}
